apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1083 @@
+---
+title: 'Python Wheels: from Tags to Variants'
+published: August 1, 2025
+authors: [michal-gorny]
+description: 'The story of how the Python Wheel Variants design was developed'
+category: [Packaging]
+featuredImage:
+  src: /posts/python-wheels-from-tags-to-variants/featured.jpeg
+  alt: A black steam train pulling a olive-colored carr in reverse. It has cool wheels with red spikes. Photograph by the author.
+hero:
+  imageSrc: /posts/python-wheels-from-tags-to-variants/hero.jpeg
+  imageAlt: A black steam train pulling a olive-colored carr in reverse. It has cool wheels with red spikes. Photograph by the author.
+---
+
+# Python Wheels: from Tags to Variants
+
+Many Python distributions are uniform across different Python versions
+and platforms. For these distributions, it is sufficient to publish
+a single binary package that can be installed everywhere. However, some
+packages are more complex than that — they include compiled Python
+extensions, binaries or even Python code that differs across systems.
+In order to robustly deploy these software on different platforms,
+you need to publish multiple binary packages, with the installers
+being able to select the one best fit the platform used.
+
+For a long time, [Python Wheels](https://packaging.python.org/en/latest/specifications/binary-distribution-format/)
+made do with a relatively simple mechanism
+of providing the needed variance: [Platform Compatibility Tags](https://packaging.python.org/en/latest/specifications/platform-compatibility-tags/).
+Tags identified different Python implementations and versions,
+operating systems, CPU architectures. Over time, they were extended
+to facilitate new use cases. To list a few: [PEP 513](https://peps.python.org/pep-0513/)
+added <code>manylinux</code> tags to standardize the dependency on GNU/Linux
+systems, [PEP 656](https://peps.python.org/pep-0656/) added
+<code>musllinux</code> tags to facilitate Linux systems with musl libc.
```

Comment:
A little pedantically, I'd say "couple" instead of "few" and use "and" instead of a comma.

Reply:
Right. I originally thought I'd list more than two, but ended up with two.